server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -459,4 +466,152 @@ protected void onDelete() {
+                captureStatsBeforeUpload();
+                uploadStartTime = System.nanoTime();
+            }
+        }
+
+        @Override
+        public void onUploadComplete(TransferSnapshot transferSnapshot) throws IOException {
+            if (toUpload != null && toUpload.size() > 0) {
+                uploadEndTime = System.nanoTime();
+                uploadEndTimeMs = System.currentTimeMillis();
+                captureStatsOnUploadSuccess();
+            }
+
+            transferReleasable.close();
+            closeFilesIfNoPendingRetentionLocks();
+            maxRemoteTranslogGenerationUploaded = generation;
+            minRemoteGenReferenced = getMinFileGeneration();
+            logger.trace("uploaded translog for {} {} ", primaryTerm, generation);
+        }
+
+        @Override
+        public void onUploadFailed(TransferSnapshot transferSnapshot, Exception ex) throws IOException {
+            if (toUpload != null && toUpload.size() > 0) {
+                uploadEndTime = System.nanoTime();
+                captureStatsOnUploadFailure();
+            }
+
+            transferReleasable.close();
+            closeFilesIfNoPendingRetentionLocks();
+            if (ex instanceof IOException) {
+                throw (IOException) ex;
+            } else {
+                throw (RuntimeException) ex;
+            }
+        }
+
+        /**
+         * Adds relevant stats to the tracker when an upload is started
+         */
+        private void captureStatsBeforeUpload() {
+            remoteTranslogTransferTracker.addUploadsStarted(toUpload.size());
+            remoteTranslogTransferTracker.addUploadBytesStarted(uploadBytes);
+        }
+
+        /**
+         * Adds relevant stats to the tracker when an upload is successfully completed
+         */
+        private void captureStatsOnUploadSuccess() {
+            long uploadDurationInMillis = (uploadEndTime - uploadStartTime) / 1_000_000L;
+            remoteTranslogTransferTracker.addUploadsSucceeded(toUpload.size());
+            remoteTranslogTransferTracker.addUploadBytesSucceeded(uploadBytes);
+            remoteTranslogTransferTracker.addUploadTimeInMillis(uploadDurationInMillis);
+            remoteTranslogTransferTracker.setLastSuccessfulUploadTimestamp(uploadEndTimeMs);
+
+            remoteTranslogTransferTracker.updateUploadBytesMovingAverage(uploadBytes);
+            remoteTranslogTransferTracker.updateUploadTimeMovingAverage(uploadDurationInMillis);
+            if (uploadDurationInMillis > 0) {
+                remoteTranslogTransferTracker.updateUploadBytesPerSecMovingAverage((uploadBytes * 1_000L) / uploadDurationInMillis);
+            }
+        }
+
+        /**
+         * Adds relevant stats to the tracker when an upload has failed
+         */
+        private void captureStatsOnUploadFailure() {
+            remoteTranslogTransferTracker.addUploadTimeInMillis((uploadEndTime - uploadStartTime) / 1_000_000L);
```

Comment:
Do we really need to add to total upload millis if there were no files uploaded? On the other side, if we are adding to total upload millis, then we can use data points for uploaded bytes and upload times for time instants and quickly find that there speed has deteriorated indicating issues with upload.

Reply:
> Do we really need to add to total upload millis if there were no files uploaded?

Sorry, I did not completely follow this. Based on my understanding, `onUploadFailed()` would be called in cases of partial failures too (i.e. for example, it would be called even if 1 out of 3 tlog files failed to get uploaded). Is `onUploadFailed()` invoked only if all uploads failed? (`onUploadFailed()` is what invokes `captureStatsOnUploadFailure()`)

Reply:
> On the other side, if we are adding to total upload millis, then we can use data points for uploaded bytes and upload times for time instants and quickly find that there speed has deteriorated indicating issues with upload.

The speed we would get here would be incorrect if we only consider the successfully uploaded bytes. Are you proposing to consider both uploaded + failed bytes in this case and then calculate the speed? If yes, we need to propagate this to segment stats and we also need to review this for other moving average stats too (currently, the moving average stats are updated only for successful refreshes/syncs).